File: trs/trs_approx.py
""" Approximate Trust Region Subproblem using Steihaug-Toint CG """
from __future__ import absolute_import, division, unicode_literals, print_function
import numpy as np
import scipy.linalg as linalg
import math as ma

""" Steihaug-Toint Conjugate Gradient """
def trs_approx(J_S, J_ST, gradf_S, delta):
    p = J_S.shape[1]

    # Parameters
    TAU = 1e-5 # Tolerance
    MAXITER = 2*p # Max iterations

    # Initialize
    s_S = np.zeros(p)
    g_S = gradf_S
    H_S_mul = lambda s: J_ST.dot(J_S.dot(s))
    ng_S2 = np.dot(g_S,g_S)
    p_S = -g_S

    k = 0
    while linalg.norm(H_S_mul(s_S) + gradf_S) > TAU*linalg.norm(gradf_S) and k < MAXITER:

        # Calculate curvature
        Jp_S = J_S.dot(p_S)
        kappa = Jp_S.dot(Jp_S)

        # Check for zero curvature
        if kappa < 1e-30: # Find boundary solution
            sigma = quadeq_pos(np.dot(p_S, p_S), 2 * np.dot(s_S, p_S), np.dot(s_S, s_S) - delta ** 2) # Find quadratic root
            return s_S + sigma * p_S

        # Calculate step length for s and g
        alpha = ng_S2/kappa

        # Trust region active: boundary solution
        if linalg.norm(s_S + alpha*p_S) >= delta:
            sigma = quadeq_pos(np.dot(p_S, p_S), 2 * np.dot(s_S, p_S), np.dot(s_S, s_S) - delta ** 2) # Find quadratic root
            return s_S + sigma * p_S

        # Take step for s and g
        s_S = s_S + alpha * p_S
        g_S = g_S + alpha * H_S_mul(p_S)

        # Calculate step length for p
        ng_S2_new = np.dot(g_S,g_S)
        beta = ng_S2_new/ng_S2
        ng_S2 = ng_S2_new

        # Take step for p
        p_S = -g_S + beta * p_S

        # Update iteration count
        k += 1

    # Trust region inactive: interior solution (or failed to converge)
    return s_S

""" Preconditioned Steihaug-Toint Conjugate Gradient """
def trs_approx_precon(J_S, J_ST, gradf_S, delta):
    p = J_S.shape[1]

    # Parameters
    TAU = 1e-5 # Tolerance
    MAXITER = 2*p # Max iterations

    # Jacobi preconditioner
    Minv = 1. / np.asarray((J_ST.multiply(J_ST)).sum(axis=-1)).ravel()  # 1/diag(J^TJ)

    # Initialize
    s_S = np.zeros(p)
    g_S = gradf_S
    v_S = Minv * g_S
    H_S_mul = lambda s: J_ST.dot(J_S.dot(s))
    g_Sv_S = np.dot(g_S,v_S)
    p_S = -v_S

    k = 0
    while linalg.norm(H_S_mul(s_S) + gradf_S) > TAU*linalg.norm(gradf_S) and k < MAXITER:

        # Calculate curvature
        Jp_S = J_S.dot(p_S)
        kappa = Jp_S.dot(Jp_S)

        # Check for zero curvature
        if kappa < 1e-30: # Find boundary solution
            sigma = quadeq_pos(np.dot(p_S, p_S), 2 * np.dot(s_S, p_S), np.dot(s_S, s_S) - delta ** 2) # Find quadratic root
            return s_S + sigma * p_S

        # Calculate step length for s and g
        alpha = g_Sv_S/kappa

        # Trust region active: boundary solution
        if linalg.norm(s_S + alpha*p_S) >= delta:
            sigma = quadeq_pos(np.dot(p_S, p_S), 2 * np.dot(s_S, p_S), np.dot(s_S, s_S) - delta ** 2) # Find quadratic root
            return s_S + sigma * p_S

        # Take step for s and g
        s_S = s_S + alpha * p_S
        g_S = g_S + alpha * H_S_mul(p_S)
        v_S = Minv * g_S

        # Calculate step length for p
        g_Sv_S_new = np.dot(g_S,v_S)
        beta = g_Sv_S_new/g_Sv_S
        g_Sv_S = g_Sv_S_new

        # Take step for p
        p_S = -v_S + beta * p_S

        # Update iteration count
        k += 1

    # Trust region inactive: interior solution (or failed to converge)
    return s_S

""" Return positive root of quadratic equation """
def quadeq_pos(a, b, c):
    x1 = np.divide(-b + ma.sqrt(b * b - 4 * a * c), 2 * a)
    x2 = np.divide(-b - ma.sqrt(b * b - 4 * a * c), 2 * a)
    return max(0,x1,x2)

File: Matlab Port/trs_approx.m
% Steihaug-Toint Conjugate Gradient
function s = trs_approx(J, gradf, delta)
    [~,n] = size(J);

    % Parameters
    TAU = 1e-5; % Tolerance
    MAXITER = 2*n; % Max iterations

    % Initialize
    s = zeros(n,1);
    g = gradf;
    ng_2 = norm(g).^2;
    p = -g;

    k = 0;
    while norm(J'*(J*s) + gradf) > TAU*norm(gradf) && k < MAXITER

        % Calculate curvature
        kappa = p'*(J'*(J*p));

        % Check for zero curvature
        if kappa < 1e-30 % Find boundary solution
            sigma = roots([p'*p, 2*s'*p, s'*s-delta^2]); % Find quadratic roots
            if isempty(sigma)
                sigma = zeros(2,1); % failed step: delta too large
            end
            s = s + max(0,max(sigma(1),sigma(2))) * p; % take positive root
            return;
        end
            
        % Calculate step length for s and g
        alpha = ng_2/kappa;

        % Trust region active: boundary solution
        if norm(s + alpha*p) >= delta
            sigma = roots([p'*p, 2*s'*p, s'*s-delta^2]); % Find quadratic roots
            if isempty(sigma)
                sigma = zeros(2,1); % failed step: delta too large
            end
            s = s + max(0,max(sigma(1),sigma(2))) * p; % take positive root
            return;
        end

        % Take step for s and g
        s = s + alpha * p;
        g = g + alpha * J'*(J*p);

        % Calculate step length for p
        ng_2_new = norm(g).^2;
        beta = ng_2_new/ng_2;
        ng_2 = ng_2_new;

        % Take step for p
        p = -g + beta * p;

        % Update iteration count
        k = k + 1;
        
    end
    
    % Trust region inactive: interior solution (or failed to converge)
   
end